src/snapshot/src/lib.rs

```diff
@@ -158,432 +160,215 @@ impl Snapshot {
-            field1: 1,
-        };
-
-        let mut snapshot_mem = vec![0u8; 1024];
-
-        // Serialize as v1.
-        let mut snapshot = Snapshot::new(vm.clone(), 1);
+        let snapshot = Snapshot::new(Version::new(12, 1, 3));
+        snapshot.save(&mut data.as_mut_slice(), &42u8).unwrap();
+
+        // Tamper the bytes written, without touching the previously CRC.
         snapshot
-            .save(&mut snapshot_mem.as_mut_slice(), &state_1)
+            .save_without_crc(&mut data.as_mut_slice(), &43u8)
             .unwrap();
-        snapshot_mem[20] = 123;
 
-        #[cfg(target_arch = "aarch64")]
-        let expected_err = Error::Crc64(0x1960_4E6A_A13F_6615);
-        #[cfg(target_arch = "x86_64")]
-        let expected_err = Error::Crc64(0x103F_8F52_8F51_20B1);
-
-        let load_result: Result<(Test1, _), Error> =
-            Snapshot::load(&mut snapshot_mem.as_slice(), 38, vm);
-        assert_eq!(load_result.unwrap_err(), expected_err);
+        assert!(matches!(
+            snapshot.load_with_version_check::<_, u8>(&mut data.as_slice(), data.len()),
+            Err(Error::Crc64(_))
+        ));
     }
 
-    #[allow(non_upper_case_globals)]
-    #[allow(non_camel_case_types)]
-    #[allow(non_snake_case)]
     #[test]
-    fn test_kvm_bindings_struct() {
-        #[repr(C)]
-        #[derive(Debug, PartialEq, Eq, Versionize)]
-        pub struct kvm_pit_config {
-            pub flags: ::std::os::raw::c_uint,
-            pub pad: [::std::os::raw::c_uint; 15usize],
-        }
-
-        let state = kvm_pit_config {
-            flags: 123_456,
-            pad: [0; 15usize],
-        };
-
-        let vm = VersionMap::new();
-        let mut snapshot_mem = vec![0u8; 1024];
-        // Serialize as v1.
-        let mut snapshot = Snapshot::new(vm.clone(), 1);
+    fn test_bad_version() {
+        let mut data = vec![0u8; 100];
+
+        // We write a snapshot with version "v1.3.12"
+        let snapshot = Snapshot::new(Version::new(1, 3, 12));
+        snapshot.save(&mut data.as_mut_slice(), &42u8).unwrap();
+
+        // Different major versions should not work
+        let snapshot = Snapshot::new(Version::new(2, 3, 12));
+        assert!(matches!(
+            snapshot.load_with_version_check::<_, u8>(&mut data.as_slice(), data.len()),
+            Err(Error::InvalidFormatVersion(Version {
+                major: 1,
+                minor: 3,
+                patch: 12,
+                ..
+            }))
+        ));
+        let snapshot = Snapshot::new(Version::new(0, 3, 12));
+        assert!(matches!(
+            snapshot.load_with_version_check::<_, u8>(&mut data.as_slice(), data.len()),
+            Err(Error::InvalidFormatVersion(Version {
+                major: 1,
+                minor: 3,
+                patch: 12,
+                ..
+            }))
+        ));
+
+        // We can't support minor versions bigger than ours
+        let snapshot = Snapshot::new(Version::new(1, 2, 12));
```

Comment:
Shouldn't this be `1.4.12` that is not supported because it is bigger than `1.3.12`?

Reply:
`1.2.12` here represents the snapshot version that a Firecracker binary supports. `1.3.12` is the version of the saved snapshot.

Reply:
ok, I get it now.